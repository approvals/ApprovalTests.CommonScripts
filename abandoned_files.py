from enum import Enum
from pathlib import Path
from typing import Callable

class Mode(Enum):
    DELETE_WITHOUT_PROMPTING = 1
    PROMPT = 2
    DRY_RUN = 3


def remove_abandoned_files(*, mode, load_touched_files, get_all_approved_files, delete,
                           system_out: Callable[[str], None] = print,
                           ):

    touched_files = load_touched_files()
    all_approved_files = get_all_approved_files()

    stray_files = [Path(file) for file in all_approved_files if file not in touched_files]
    for stray_file in stray_files:
        delete(str(stray_file))
    report(system_out, stray_files)


def report(system_out, stray_files):
    system_out("Unused `.approved.` files found.\n")

    for stray_file in stray_files:
        system_out(f" - {stray_file.name} (in {stray_file.parent.as_posix()}/)")

    system_out("")
    system_out(f"Deleted {len(stray_files)} files.")
