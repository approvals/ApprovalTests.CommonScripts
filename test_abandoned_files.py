from abandoned_files import remove_abandoned_files, Mode
from approvaltests import verify, Options
from approvaltests.reporters import GenericDiffReporter, GenericDiffReporterConfig


def test__find_abandoned_files__with_loader_and_saver():
    def load_touched_files():
        return ["b.approved.txt"]

    def get_all_approved_files():
        return ["a.approved.txt", "b.approved.txt"]

    deletes = []

    def delete(file):
        nonlocal deletes
        deletes.append(file.name)

    remove_abandoned_files(
        mode=Mode.DELETE_WITHOUT_PROMPTING,
        load_touched_files=load_touched_files,
        get_all_approved_files=get_all_approved_files,
        delete=delete,
    )

    assert deletes == ["a.approved.txt"]


def test__console_output():
    verify_abandoned_files(
        [
            "path/to/a.stray.approved.txt",
            "b.approved.txt",
            "path/to2/c.stray.approved.txt",
        ]
    )


def verify_abandoned_files(files):
    def load_touched_files():
        return [file for file in files if "stray" not in file]

    def get_all_approved_files():
        return files

    def delete(file):
        pass

    result = ""

    def system_out(text):
        nonlocal result
        result += text + "\n"

    remove_abandoned_files(
        mode=Mode.DELETE_WITHOUT_PROMPTING,
        load_touched_files=load_touched_files,
        get_all_approved_files=get_all_approved_files,
        delete=delete,
        system_out=system_out,
    )

    class ReportWithBeyondCompare5Windows(GenericDiffReporter):
        def __init__(self):
            super().__init__(
                config=GenericDiffReporterConfig(
                    name=self.__class__.__name__,
                    path="{ProgramFiles}/Beyond Compare 5/BCompare.exe",
                )
            )

    verify(result, options=Options().with_reporter(ReportWithBeyondCompare5Windows()))
