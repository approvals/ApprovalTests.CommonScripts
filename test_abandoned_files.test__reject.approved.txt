Unused `.approved.` files found.

 - a.stray.approved.txt (in path/to/)
 - c.stray.approved.txt (in path/to2/)
Delete? [Y/n]
n
No files were deleted.
